docs(zh): translate intergration and vs (#549)
withastro/docs: src/pages/zh-CN/comparing-astro-vs-other-tools.md

---
layout: ~/layouts/MainLayout.astro
title: Astro vs. X
description: 比较 Astro 和诸如 Gatsby、Next.js、Nuxt、Hugo、Eleventy 之类的静态网站生成器。
---
<!-- TODO: UNcomment out the parts re: number of bytes of JS etc, once we decide which values/markers we'd like to use here. -->

> 我们经常会被问道这样一个问题“Astro 和我最喜欢的 **\_\_\_\_** 项目比起来怎么样？”

Reviewer: 我们经常会被问到这样一个问题” Astro 和我最喜欢的 **\_\_\_\_** 项目比起来怎么样？”

Author: Ok, a typo

> 我们经常会被问道这样一个问题“Astro 和我最喜欢的 **\_\_\_\_** 项目比起来怎么样？”

Reviewer: ```suggestion
我们经常会被问到这样一个问题“Astro 和我最喜欢的 **\_\_\_\_** 项目比起来怎么样？”
```

本指南就是为了回答这个问题而编写的，适用于几个流行的网站生成器和 Astro 替代品。

> 两个重要特性使得 Astro 与大多数替代品不同。

Reviewer: 两个重要特性使得 Astro 与大多数替代品不同：

> 两个重要特性使得 Astro 与大多数替代品不同。

Reviewer: ```suggestion
两个重要特性使得 Astro 与大多数替代品不同：
```

> - [按需加载](/en/core-concepts/partial-hydration)

Reviewer: ```suggestion
- [极简耦合](/zh-CN/core-concepts/partial-hydration/)
```

Author: What meaning of `偏`? `按需加载` is used here instead of `部分激活` because I got a better view of the translation of the word halfway through the translation of the document. Here's the part I accidentally left out.

> - [使用你喜欢的框架](/en/core-concepts/framework-components)

Reviewer: - [极简耦合](/zh-CN/core-concepts/partial-hydration/)
- [框架组件](/zh-CN/core-concepts/framework-components/)

> - [使用你喜欢的框架](/en/core-concepts/framework-components)

Reviewer: ```suggestion
- [框架组件](/zh-CN/core-concepts/framework-components/)
```

你可以在这个页面上通过我们的深入比较了解到更多细节。

如果你最喜欢的生成器没有列在下方，你可以 [在 Discord 里问我们](https://astro.build/chat)。

## Docusaurus vs. Astro

[Docusaurus](https://docusaurus.io/) 是一个流行的文档网站生成器。Docusaurus 使用 React 来生成网站 UI，而 Astro 支持 React、Preact、Vue.js、Svelte、SolidJS、AlpineJS、Lit 和原始 HTML 模板。

> Docusaurus 是为建立文档网站而设计的，它有一些内置的、针对文档的网站功能，而 Astro 没有。不够 Astro 通过官方的 [`docs`](https://github.com/withastro/astro/tree/main/examples/docs) 主题提供了特定的文档功能，你可以在自己的网站上使用。这个网站正是用这个模板建立的！

Reviewer: Docusaurus 是为建立文档网站而设计的，它有一些内置的、而 Astro 没有的特定于文档的网站功能。不过， Astro 通过官方的 [`docs`](https://github.com/withastro/astro/tree/main/examples/docs) 主题提供了特定于文档功能，你可以在自己的网站上使用它。这个网站正是用这个模板建立的！

> Docusaurus 是为建立文档网站而设计的，它有一些内置的、针对文档的网站功能，而 Astro 没有。不够 Astro 通过官方的 [`docs`](https://github.com/withastro/astro/tree/main/examples/docs) 主题提供了特定的文档功能，你可以在自己的网站上使用。这个网站正是用这个模板建立的！

Reviewer: ```suggestion
Docusaurus 是为建立文档网站而设计的，它有一些内置的、而 Astro 没有的特定于文档的网站功能。不过， Astro 通过官方的 [docs](https://github.com/withastro/astro/tree/main/examples/docs) 主题提供了特定于文档功能，你可以在自己的网站上使用它。这个网站正是用这个模板建立的！
```

#### 比较 Docusaurus vs. Astro 性能

> 在大多数情况下，Astro 网站的加载速度将明显快于 Docusaurus 网站。这是因为 Astro 会自动从页面中剥离不必要的 JavaScript，只激活个别需要的组件。这个功能被称为[部分激活](/en/core-concepts/partial-hydration)。

Reviewer: ```suggestion
在大多数情况下，Astro 网站的加载速度将明显快于 Docusaurus 网站。这是因为 Astro 会自动从页面中剥离不必要的 JavaScript，只激活个别需要的组件。这个功能被称为[极简耦合](https://github.com/zh-CN/core-concepts/partial-hydration/)。
```

> Docusaurus 不支持部分激活，而是让用户在浏览器中加载并重新激活整个页面，即便大部分的页面内容都是静态的，这也使得网站页面加载比较慢和性能比较差。在 Docusaurus 中没有办法禁用这种行为。

Reviewer: ```suggestion
Docusaurus 不支持极简耦合，而是让用户在浏览器中加载并重新激活整个页面，即便大部分的页面内容都是静态的，这也使得网站页面加载比较慢和性能比较差。在 Docusaurus 中没有办法禁用这种行为。
```

#### 案例：创建文档站点

[docusaurus.io/docs](https://docusaurus.io/docs) 是 Docusaurus 官方文档网站，不言而喻它自然是用 Docusaurus 构建的。与 Astro 官方文档网站相比较，二者的设计和功能十分相似。下面给出了两个站点生成器的**粗略但又真实**的比较。

> - **Docusaurus 性能得分**：53 分（满分 100 分）[（全面检查）](/lighthouse/docusaurus/)
> - **Astro 性能得分**：92 分（满分 100 分）[（全面检查）](/lighthouse/astro/)

Reviewer: ```suggestion
- **Docusaurus 性能得分**：53 分（满分 100 分）[（检查全文）](/lighthouse/docusaurus/)
- **Astro 性能得分**：92 分（满分 100 分）[（检查全文）](/lighthouse/astro/)
```

<!-- One big reason behind this performance difference is Astro’s smaller JavaScript payload: [docusaurus.io/docs](https://docusaurus.io/docs) loads **238kb** of JavaScript on first page load while [docs.astro.build](https://docs.astro.build) loads **78.7kb** (67% less JavaScript, overall) _after_ first load.) -->

## Elder.js vs. Astro

[Elder.js](https://elderguide.com/tech/elderjs/) 是一个有态度、为 Svelte 打造的静态网站生成器。

> Elder.js 使用 Svelte 来渲染网站内容。Astro 更灵活：你可以自由地使用任何流行的组件库（React、Preact、Vue、Svelte、Solid 和其他）或 Astro 的类似 HTML 的组件语法来构建 UI，这类似于 HTML + JSX。

Reviewer: ```suggestion
Elder.js 使用 Svelte 来渲染网站内容。相比较，Astro 更灵活：你可以自由地使用任何流行的组件库（React、Preact、Vue、Svelte、Solid 和其他）或 与 HTML + JSX 相似的Astro 类 HTML 的组件语法来构建 UI。
```

> Elder.js 在这个列表中是独一无二的，因为它是唯一个支持[部分激活](/en/core-concepts/partial-hydration)的生成器。Astro 和 Elder.js 都会自动从页面中剥离不必要的 JavaScript，只激活个别需要的组件。Astro 与 Elder 的部分激活在 API 方面有点不同，Astro 支持一些 Elder.js 没有的功能（如 `client:media`)。然而就性能而言，这两个项目构建的网站基本一致。

Reviewer: ```suggestion
Elder.js 在这个列表中是独一无二的，因为它是支持[极简耦合](/zh-CN/core-concepts/partial-hydration)的唯一的一个其他网站生成器。Astro 和 Elder.js 都会自动从页面中剥离不必要的 JavaScript，只激活个别需要的组件。Astro 与 Elder 的部分激活在 API 方面有点不同，Astro 支持一些 Elder.js 没有的功能（如 `client:media`)。然而就性能而言，这两个项目构建的网站基本一致。
```

> Elder.js 使用自定义的路由解决方案，对于新的开发者来说可能会感到陌生。而 Astro 则使用[基于文件的路由](/en/core-concepts/routing)，这对任何来自 Next.js\SvelteKit，甚至其他静态网站生成器如 Eleventy 的人来说都应该感到熟悉。

Reviewer: ```suggestion
Elder.js 使用自定义的路由解决方案，对于新的开发者来说可能会感到陌生。而 Astro 则使用[基于文件的路由](/zh-CN/core-concepts/routing)，这对任何来自 Next.js\SvelteKit，甚至其他静态网站生成器如 Eleventy 的人来说都应该感到熟悉。
```

Author: This page has not been translated, so it should not be changed.

I plan to do a full revision when all pages are translated.

Reviewer 2: We should use `/zh-CN/` links here! That way we don’t need to go back and update them later — the page already exists: https://docs.astro.build/zh-CN/core-concepts/routing/

Author: Have I translated it yet? I didn't pay attention to it.

But I still prefer to update it with the last, otherwise every time I submit a translation I still need to check if there is a link to this page in the previous translation, and then modify it.

Author: It's will make ci error, right? Should we ignore it.

Reviewer 2: No, it will pass. The CI checks the actual built URLs rather than which `.md` files exist.

Elder.js 被设计用于大型网站，并声称可以在不到十分钟的时间内（在一个普通的虚拟机上）建立一个约两万个页面的网站。在写这篇文章的时候，Astro 在 66 秒内构建约一千个页面，但还没有在有两千个以上页面的项目上测试过。

> Elder.js 同时支持静态网站生成（SSG）和服务器端渲染（SSR）。Astro 可以用 SSG 静态构建或用[适配器](/en/guides/server-side-rendering/#enabling-ssr-in-your-project)部署到 SSR 环境，像是 Deno、Vercel serverless、Netlify serverless 和 Node.js，并将支持更多服务。

Reviewer: ```suggestion
Elder.js 同时支持静态网站生成（SSG）和服务器端渲染（SSR）。Astro 可以用 SSG 静态构建或用像 Deno、Vercel serverless、Netlify serverless 和 Node.js 之类的[适配器](/en/guides/server-side-rendering/#enabling-ssr-in-your-project)部署到 SSR 环境，并将支持更多的适配器。
```

## Eleventy vs. Astro

[Eleventy](https://www.11ty.dev/) 是一个由 Node.js 驱动的流行静态网站生成器。

> Eleventy 使用几种[旧的 HTML 模板语言](https://www.11ty.dev/docs/languages/) 对网站进行渲染，Nunjucks、Liquid、Pug、EJS 和其他的模板语言。Astro 则能让你使用你喜欢的 UI 组件库（React、Preact、Vue、Svelte 等）或类用似于 HTML + JSX 的内置组件语法来创建页面。Eleventy 不支持使用现代 UI 组件进行 HTML 模板制作。

Reviewer: ```suggestion
Eleventy 使用几种像 Nunjucks、Liquid、Pug、EJS 和其他的[过时的 HTML 模板语言](https://www.11ty.dev/docs/languages/) 对网站进行渲染。Astro 则能让你使用你喜欢的 UI 组件库（React、Preact、Vue、Svelte 等）或类似于 HTML + JSX 的内置组件语法来创建页面。Eleventy 不支持使用现代 UI 组件进行 HTML 模板制作。
```

Author: old doesn't mean outdated

#### 比较 Eleventy vs. Astro 性能

从概念上讲，Eleventy 和 Astro 的“最小客户端 JavaScript”的网络开发方法是一致的。Eleventy 和 Astro 都提供类似的零 JavaScript 默认性能基准。

Eleventy 通过推动你完全避免使用 JavaScript 来实现这一目标。Eleventy 的网站通常很少或根本没有使用 JavaScript。当你确实需要使用客户端 JavaScript 时，这就成了一个问题。你可以自行为 Eleventy 创建资源构建 pipeline。这可能会耗时很久，并使得你不得不自己在捆绑、压缩和其他方面上进行复杂配置以实现优化。

> 相比之下，Astro 自动为你构建客户端 JavaScript 和 CSS。Astro 自动从页面中剥离不必要的 JavaScript，只激活个别需要的组件。这个功能被称为[部分激活](/en/core-concepts/partial-hydration)。虽然在 Eleventy 中可以自己实现这个功能，但 Astro 内置了该功能。

Reviewer: ```suggestion
相比之下，Astro 自动为你构建客户端 JavaScript 和 CSS。Astro 自动从页面中剥离不必要的 JavaScript，只激活个别需要的组件。这个功能被称为[极简耦合](/zh-CN/core-concepts/partial-hydration)。虽然在 Eleventy 中可以自己实现这个功能，但 Astro 内置了该功能。
```

#### 案例：创建文档站点

> [11ty.dev/docs](https://www.11ty.dev/docs/) 是 11ty 的官方文档网站，由 11ty 建立。与 Astro 官方文档网站相比较，二者的设计和功能十分相似。下面面给出了两个站点生成器的**粗略但又真实**的比较。
> 
> - **11ty 性能得分**：86 分（满分 100 分）[（全面检查）](/lighthouse/11ty/)
> - **Astro 性能得分**：92 分（满分 100 分）[（全面检查）](/lighthouse/astro/)

Reviewer: ```suggestion
[11ty.dev/docs](https://www.11ty.dev/docs/) 是 11ty 的官方文档网站，由 11ty 建立。与 Astro 官方文档网站相比较，二者的设计和功能十分相似。下面给出了两个站点生成器**粗略但又真实**的比较。

- **11ty 性能得分**：86 分（满分 100 分）[（检查全文）](/lighthouse/11ty/)
- **Astro 性能得分**：92 分（满分 100 分）[（检查全文）](/lighthouse/astro/)
```

Author: It doesn't provide words, it's just data. `全文` (full text) is not correct

## Gatsby vs. Astro

[Gatsby](https://www.gatsbyjs.com/) 是一个流行的 React 的网站和应用程序框架。

> Gatsby 使用 React 来渲染网站内容。Astro 更加灵活：你可以自由地使用任何流行的组件库（React、Preact、Vue、Svelte、Solid 和其他）或 Astro 的类似 HTML 的组件语法来构建 UI，这与 HTML+JSX 相似。

Reviewer: ```suggestion
Gatsby 使用 React 来渲染网站内容。Astro 更加灵活：你可以自由地使用任何流行的组件库（React、Preact、Vue、Svelte、Solid 和其他）或 与 HTML+JSX 相似的Astro 类 HTML 组件语法来构建 UI。
```

> Gatsby v4 同时支持静态网站生成（SSG）、增量重建、延迟静态生成（DSG）和服务器端渲染（SSR）。Astro 可以用 SSG 静态构建或用[适配器](/en/guides/server-side-rendering/#enabling-ssr-in-your-project)部署到 SSR 环境，像是 Deno、Vercel serverless、Netlify serverless 和 Node.js，并将支持更多服务。

Reviewer: ```suggestion
Gatsby v4 支持具有增量重建的静态站点生成 (SSG)、延迟静态生成 (DSG) 和服务器端渲染 (SSR)。Astro 可以用 SSG 静态构建或用 Deno、Vercel serverless、Netlify serverless 和 Node.js之类的[适配器](/en/guides/server-side-rendering/#enabling-ssr-in-your-project)部署到 SSR 环境，并将支持更多的适配器。
```

> Gatsby 需要定制一个 GraphQL API 来处理你所有的网站内容。虽然有些开发者喜欢这种模式，但经常有对 Gatsby 的批评，认为这种模式会随着时间的推移和网站的增长，变得过于复杂和难以维护。Astro 没有对 GraphQL 需求，它提供熟悉的 API（如 `fetch ()` 和顶层的 `await`），以便在需要数据的地方加载数据。然而你可以随意选择将 Astro 和任何服务器端或客户端的 GraphQL 库一起使用。

Reviewer: ```suggestion
Gatsby 需要定制一个 GraphQL API 来处理你所有的网站内容。虽然有些开发者喜欢这种模式，但经常有对 Gatsby 的批评，认为这种模式会随着时间的推移和网站的增长，变得过于复杂和难以维护。Astro 没有对 GraphQL 的需求，它提供熟悉的 API（如 `fetch ()` 和顶层的 `await`），以便在需要数据的地方加载数据。然而你可以随意选择将 Astro 和任何服务端或客户端的 GraphQL 库一起使用。
```

#### 比较 Gatsby vs. Astro 性能

> 在大多数情况下，Astro 网站的加载速度会明显快于 Gatsby 网站。这是因为 Astro 会自动将不必要的 JavaScript 从页面中剥离出来，只激活个别需要的组件。这个功能被称为[部分激活](/en/core-concepts/partial-hydration)。

Reviewer: ```suggestion
在大多数情况下，Astro 网站的加载速度会明显快于 Gatsby 网站。这是因为 Astro 会自动将不必要的 JavaScript 从页面中剥离出来，只激活个别需要的组件。这个功能被称为[极简耦合](/zh-CN/core-concepts/partial-hydration)。
```

Gatsby 不支持部分激活，而是让用户在浏览器中加载和重新激活整个页面，即便大部分的页面内容都是静态的，这也使得网站页面加载比较慢和性能比较差。Gatsby 有用于移除页面中的所有 JavaScript 的[社区插件](https://www.gatsbyjs.com/plugins/gatsby-plugin-no-javascript/)，但这将破坏许多网站。这使得你不得不在在每个页面的交互性上做出一个全有或全无的决定。

> Gatsby 有个伟大的插件生态系统，而 Astro 的[集成集合](https://astro.build/integrations/)刚刚起步但仍在不断成长。Gatsby 提供了用于高级图像优化的 [gatsby-plugin-image](https://www.gatsbyjs.com/plugins/gatsby-plugin-image/)。虽然 Astro 并没有提供没有类似的第一方解决方案，但流行的社区集成 [astro-imagetools](https://github.com/RafidMuhymin/astro-imagetools#readme) 可以用于优化图像、背景图像和生成响应式图像。

Reviewer: ```suggestion
Gatsby 拥有出色的插件生态系统，而 Astro 的[集成集合](https://astro.build/integrations/)刚刚起步但仍在不断增长。Gatsby 提供了用于高级图像优化的 [gatsby-plugin-image](https://www.gatsbyjs.com/plugins/gatsby-plugin-image/)。虽然 Astro 并没有提供类似可比的第一方解决方案，但流行的社区集成 [astro-imagetools](https://github.com/RafidMuhymin/astro-imagetools#readme) 可以用于优化图像、背景图像和生成响应式图像。
```

#### 案例：创建文档站点

[gatsbyjs.com/docs](https://www.gatsbyjs.com/docs/quick-start/) 是 Gatsby 的官方文档网站，用 Gatsby 构建。与 Astro 官方文档网站相比较，二者的设计和功能十分相似。下面给出了两个站点生成器的**粗略但又真实**的比较。

> - **Gatsby 性能得分**：46 分（满分 100 分）[（全面检查）](/lighthouse/gatsby/)
> - **Astro 性能得分**：92 分（满分 100 分）[（全面检查）](/lighthouse/astro/)

Reviewer: ```suggestion
- **Gatsby 性能得分**：46 分（满分 100 分）[（检查全文）](/lighthouse/gatsby/)
- **Astro 性能得分**：92 分（满分 100 分）[（检查全文）](/lighthouse/astro/)
```

<!-- One big reason behind this performance difference is Astro’s smaller JavaScript payload: [gatsbyjs.com/docs](https://www.gatsbyjs.com/docs/quick-start/) loads **417kb** of JavaScript on first page load while [docs.astro.build](https://docs.astro.build) loads **78.7kb** (81% less JavaScript, overall) _after_ first load. -->

## Hugo vs. Astro

[Hugo](https://gohugo.io/) 是一个由 Go 驱动的流行静态网站生成器。

Hugo 使用自定义的[模板语言](https://gohugo.io/templates/introduction/)来渲染网站内容。Astro 让你使用你喜欢的 UI 组件库（React、Preact、Vue、Svelte 等）或类似于 HTML + JSX 的内置组件语法创建页面。Hugo 不支持使用现代 UI 组件进行 HTML 模板制作。

#### 比较 Hugo vs. Astro 性能

在概念上，Hugo 与 Astro 的“最小客户端 JavaScript”的网络开发方法是一致的。Hugo 和 Astro 都提供了类似的零 JavaScript 默认的性能基线。

> Hugo 和 Astro 都提供了对构建、捆绑和最小化 JavaScript 的内置支持。Astro 会自动从页面中剥离不必要的 JavaScript，只激活个别需要的组件。这个功能被称为[部分激活](/en/core-concepts/partial-hydration)。虽然在 Hugo 中可以自己实现这个功能，但 Astro 内置了该功能。

Reviewer: ```suggestion
Hugo 和 Astro 都提供了对构建、捆绑和最小化 JavaScript 的内置支持。Astro 会自动从页面中剥离不必要的 JavaScript，只激活个别需要的组件。这个功能被称为[极简耦合](/zh-CN/core-concepts/partial-hydration)。虽然在 Hugo 中可以自己实现这个功能，但 Astro 内置了该功能。
```

#### 案例：创建文档站点

[gohugo.io/documentation/](https://gohugo.io/documentation/) 是 Hugo 的官方文档网站，用 Hugo 构建。与 Astro 官方文档网站相比较，二者的设计和功能十分相似。下面给出了两个站点生成器的**粗略但又真实**的比较。

> - **Hugo 性能得分**：98 分（满分 100 分）[（全面检查）](/lighthouse/hugo/)
> - **Astro 性能得分**：92 分（满分 100 分）[（全面检查）](/lighthouse/astro/)

Reviewer: ```suggestion
- **Hugo 性能得分**：98 分（满分 100 分）[（检查全文）](/lighthouse/hugo/)
- **Astro 性能得分**：92 分（满分 100 分）[（检查全文）](/lighthouse/astro/)
```

| ---------------- | --------------------------------------------------- | ----------------------------------------------------- |
| 文件扩展名       | `.astro`                                            | `.jsx` 或 `.tsx`                                      |
| 用户定义的组件   | `<Capitalized>`                                     | `<Capitalized>`                                       |
| 表达式语法       | `{}`                                                | `{}`                                                  |
| 展开属性         | `{...props}`                                        | `{...props}`                                          |
| 布尔属性         | `autocomplete` === `autocomplete={true}`            | `autocomplete` === `autocomplete={true}`              |
| 行内函数         | `{items.map (item => <li>{item}</li>)}`             | `{items.map (item => <li>{item}</li>)}`               |
| 条件渲染         | `{condition &&  <p>text<p>}`                        | `{condition &&  <p>text<p>}`                          |
| IDE 支持         | [VS Code（包括 Open VSX）和 Nova](/en/editor-setup) | Phenomenal                                            |

> | 需要在 JS 中导入 | No                                                  | Yes，比如在作用域内使用 `jsxPragma`（`React` 或 `h`） |

Reviewer: ```suggestion
| 需要JS 导入 | No                                                  | Yes，比如在作用域内使用 `jsxPragma`（`React` 或 `h`） |
```

| 片段             | 在顶级自动，在函数中使用 `<Fragment>` 或 `<>`       | 使用 `<Fragment>` 或 `<>` 包装                        |
| 单文件多框架     | Yes                                                 | No                                                    |
| 修改 `<head>`    | 只需要在顶级页面使用 `<head>`                       | 每个框架不同（如 `<Head>`, `<svelte:head>`）          |

> | 评论风格         | `<!-- HTML -->`                                     | `{/_ JavaScript _/}`                                  |

Reviewer: ```suggestion
| 注释风格         | `<!-- HTML -->`                                     | `{/_ JavaScript _/}`                                  |
```
